security/advancedtls/pemfile_provider_test.go

```diff
@@ -0,0 +1,216 @@
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.desc, func(t *testing.T) {
+			provider, err := NewPEMFileProvider(test.options)
+			if (err != nil) != test.wantError {
+				t.Fatalf("NewPEMFileProvider(%v) = %v, want %v", test.options, err, test.wantError)
+			}
+			if err != nil {
+				return
+			}
+			provider.Close()
+		})
+	}
+
+}
+
+// This test overwrites the credential reading function used by the watching
+// goroutine. It is tested under different stages:
+// At stage 0, we force reading function to load clientPeer1 and serverTrust1,
+// and see if the credentials are picked up by the watching go routine.
+// At stage 1, we force reading function to cause an error. The watching go
+// routine should log the error while leaving the credentials unchanged.
+// At stage 2, we force reading function to load clientPeer2 and serverTrust2,
+// and see if the new credentials are picked up.
+func (s) TestWatchingRoutineUpdates(t *testing.T) {
+	// Load certificates.
+	cs := &certStore{}
+	if err := cs.loadCerts(); err != nil {
+		t.Fatalf("cs.loadCerts() failed: %v", err)
+	}
+	tests := []struct {
+		desc         string
+		options      PEMFileProviderOptions
+		wantKmStage0 certprovider.KeyMaterial
+		wantKmStage1 certprovider.KeyMaterial
+		wantKmStage2 certprovider.KeyMaterial
+	}{
+		{
+			desc: "use identity certs and root certs",
+			options: PEMFileProviderOptions{
+				CertFile:  "not_empty_cert_file",
+				KeyFile:   "not_empty_key_file",
+				TrustFile: "not_empty_trust_file",
+			},
+			wantKmStage0: certprovider.KeyMaterial{Certs: []tls.Certificate{cs.clientPeer1}, Roots: cs.serverTrust1},
+			wantKmStage1: certprovider.KeyMaterial{Certs: []tls.Certificate{cs.clientPeer1}, Roots: cs.serverTrust1},
+			wantKmStage2: certprovider.KeyMaterial{Certs: []tls.Certificate{cs.clientPeer2}, Roots: cs.serverTrust2},
+		},
+		{
+			desc: "use identity certs only",
+			options: PEMFileProviderOptions{
+				CertFile: "not_empty_cert_file",
+				KeyFile:  "not_empty_key_file",
+			},
+			wantKmStage0: certprovider.KeyMaterial{Certs: []tls.Certificate{cs.clientPeer1}},
+			wantKmStage1: certprovider.KeyMaterial{Certs: []tls.Certificate{cs.clientPeer1}},
+			wantKmStage2: certprovider.KeyMaterial{Certs: []tls.Certificate{cs.clientPeer2}},
+		},
+		{
+			desc: "use trust certs only",
+			options: PEMFileProviderOptions{
+				TrustFile: "not_empty_trust_file",
+			},
+			wantKmStage0: certprovider.KeyMaterial{Roots: cs.serverTrust1},
+			wantKmStage1: certprovider.KeyMaterial{Roots: cs.serverTrust1},
+			wantKmStage2: certprovider.KeyMaterial{Roots: cs.serverTrust2},
+		},
+	}
+	for _, test := range tests {
+		testInterval := 200 * time.Millisecond
+		test.options.IdentityInterval = testInterval
+		test.options.RootInterval = testInterval
+		t.Run(test.desc, func(t *testing.T) {
+			stage := &stageInfo{}
+			oldReadKeyCertPairFunc := readKeyCertPairFunc
+			readKeyCertPairFunc = func(certFile, keyFile string) (tls.Certificate, error) {
```

Comment:
Another option here is to use https://golang.org/pkg/io/ioutil/#TempFile, and populate them with actual contents, so that you dont have to override the `readKeyCertPairFunc` and `readTrustCertFunc`. That way you will also be exercising the functionality in `readTrustCertFunc`. This need not be done as part of this PR. But if you feel this is a good idea, then you can add a TODO and do it in a later PR.
With this approach, you would not need any of this `stage` business. You would write actual contents into the temp files from your test, and pass the tempFile path to the code which will read directly from there.

Reply:
Yes, actually we originally used this TempFile in https://github.com/grpc/grpc-go/pull/3785, but later changed to this implementation because we think making changes to local file system(even though it's in tmp folder) might be too much for unit tests.
I might use this when writing the integration tests, or when creating examples for the whole dynamic loading feature. Anyway thanks for the suggestions!